datafusion/physical-expr/src/expressions/binary.rs

```diff
@@ -62,7 +64,27 @@ impl BinaryExpr {
         op: Operator,
         right: Arc<dyn PhysicalExpr>,
     ) -> Self {
-        Self { left, op, right }
+        Self {
+            left,
+            op,
+            right,
+            fail_on_overflow: false,
+        }
+    }
+
+    /// Create new binary expression with explicit fail_on_overflow value
```

Comment:
Another alternative API might be more of a builder style and easier to extend in the fuuure

```rust
        // create expression
        let expr = BinaryExpr::new(
            Arc::new(Column::new("l", 0)),
            Operator::Plus,
            Arc::new(Column::new("r", 1)),
         )
          // configure to fail on overflow
          .with_fail_on_overflow(true);
```